pkg/scheduler/objects/reservation.go

```diff
@@ -61,51 +60,21 @@ func newReservation(node *Node, app *Application, ask *Allocation, appBased bool
 	return res
 }
 
-func reservationKey(node *Node, app *Application, ask *Allocation) string {
-	if ask == nil || (app == nil && node == nil) || (app != nil && node != nil) {
-		log.Log(log.SchedReservation).Warn("Illegal reservation key requested",
-			zap.Any("node", node),
-			zap.Any("app", app),
-			zap.Any("ask", ask))
-		return ""
-	}
-	if node == nil {
-		return ask.resKeyWithoutNode
-	}
-	key := ask.getReservationKeyForNode(node.NodeID)
-	if key != common.Empty {
-		return key
-	}
-
-	key = node.NodeID + "|" + ask.GetAllocationKey()
-	ask.setReservationKeyForNode(node.NodeID, key)
-	return key
-}
-
-func reservationKeyWithoutNode(appID, allocKey string) string {
-	return appID + "|" + allocKey
-}
-
-// Return the reservation key
-func (r *reservation) getKey() string {
-	if r.nodeID == "" {
-		return r.appID + "|" + r.askKey
-	}
-	return r.nodeID + "|" + r.askKey
-}
-
 func (r *reservation) String() string {
+	if r == nil {
+		return "nil reservation"
+	}
 	if r.nodeID == "" {
-		return r.node.NodeID + " -> " + r.appID + "|" + r.askKey
+		return r.node.NodeID + " -> " + r.appID + "|" + r.allocKey
 	}
-	return r.app.ApplicationID + " -> " + r.nodeID + "|" + r.askKey
+	return r.app.ApplicationID + " -> " + r.nodeID + "|" + r.allocKey
 }
```

Comment:
This is so interesting here. If `nodeID` == "" then we just take the it from `r.node.NodeID` anyway.  To me it looks like we're storing a string value for boolean purposes (we should store `appBased` instead), but it's not clear to me what happens here. Also `r.app.ApplicationID` could be just `r.appID`, correct?

Reply:
if the `nodeID` stored in the reservation is not an empty string the `appID` stored is an empty string. One of the two is always empty. Instead of testing twice we use that to only test once and return the correct details for both cases.
It is the only way we can distinguish between the allocation belonging to either the application or the node object. We could have further simplified it by using a boolean and always get the IDs directly out of the linked objects. That can still be done without a large impact on the code. Might change some logging entries...